File: Arrays/18-4Sum/Python/18-4Sum-Optimal.py
"""
Problem: 18. 4Sum
Approach: Optimal - Sorting + Two Pointers

Sort the array.
Fix the first two elements and use two pointers for
the remaining two elements.

Skip duplicate values for i, j, left, and right
to avoid duplicate quadruplets.

Time Complexity: O(n^3)
Space Complexity: O(1) auxiliary
"""

class Solution:
    def fourSum(self, nums: list[int], target: int) -> list[list[int]]:
        nums.sort()
        n = len(nums)
        result = []

        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            for j in range(i + 1, n):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue

                left = j + 1
                right = n - 1

                while left < right:
                    total = (
                        nums[i] +
                        nums[j] +
                        nums[left] +
                        nums[right]
                    )

                    if total == target:
                        result.append([
                            nums[i],
                            nums[j],
                            nums[left],
                            nums[right]
                        ])

                        left += 1
                        right -= 1

                        while left < right and nums[left] == nums[left - 1]:
                            left += 1

                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1

                    elif total < target:
                        left += 1

                    else:
                        right -= 1

        return result

File: Arrays/18-4Sum/JavaScript/18-4Sum-Optimal.js
/**
 * Problem: 18. 4Sum
 * Approach: Optimal - Sorting + Two Pointers
 *
 * Sort the array.
 * Fix the first two elements and use two pointers
 * for the remaining two elements.
 *
 * Skip duplicates at every level.
 *
 * Time Complexity: O(n^3)
 * Space Complexity: O(1) auxiliary
 */

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    const n = nums.length;
    const result = [];

    nums.sort((a, b) => a - b);

    for (let i = 0; i < n; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }

        for (let j = i + 1; j < n; j++) {
            if (j > i + 1 && nums[j] === nums[j - 1]) {
                continue;
            }

            let left = j + 1;
            let right = n - 1;

            while (left < right) {
                const totalSum =
                    nums[i] +
                    nums[j] +
                    nums[left] +
                    nums[right];

                if (totalSum === target) {
                    result.push([
                        nums[i],
                        nums[j],
                        nums[left],
                        nums[right]
                    ]);

                    left++;
                    right--;

                    while (
                        left < right &&
                        nums[left] === nums[left - 1]
                    ) {
                        left++;
                    }

                    while (
                        left < right &&
                        nums[right] === nums[right + 1]
                    ) {
                        right--;
                    }

                } else if (totalSum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
    }

    return result;
};